cmd/csi-sanity/sanity_test.go

```diff
@@ -30,38 +30,58 @@ const (
-	flag.BoolVar(&config.TestNodeVolumeAttachLimit, prefix+"testnodevolumeattachlimit", false, "Test node volume attach limit")
-	flag.StringVar(&config.JUnitFile, prefix+"junitfile", "", "JUnit XML output file where test results will be written")
-	flag.Parse()
+func stringVar(p *string, name string, usage string) {
+	flag.StringVar(p, prefix+name, *p, usage)
 }
 
-func TestSanity(t *testing.T) {
-	if version {
+func boolVar(p *bool, name string, usage string) {
+	flag.BoolVar(p, prefix+name, *p, usage)
+}
+
+func intVar(p *int, name string, usage string) {
+	flag.IntVar(p, prefix+name, *p, usage)
+}
+
+func int64Var(p *int64, name string, usage string) {
+	flag.Int64Var(p, prefix+name, *p, usage)
+}
+
+func TestMain(m *testing.M) {
+	version := flag.Bool("version", false, "print version of this program")
+
+	// Support overriding the default configuration via flags.
+	stringVar(&config.Address, "endpoint", "CSI endpoint")
+	stringVar(&config.ControllerAddress, "controllerendpoint", "CSI controller endpoint")
+	stringVar(&config.TargetPath, "mountdir", "Mount point for NodePublish")
+	stringVar(&config.StagingPath, "stagingdir", "Mount point for NodeStage if staging is supported")
+	stringVar(&config.CreateTargetPathCmd, "createmountpathcmd", "Command to run for target path creation")
+	stringVar(&config.CreateStagingPathCmd, "createstagingpathcmd", "Command to run for staging path creation")
+	intVar(&config.CreatePathCmdTimeout, "createpathcmdtimeout", "Timeout for the commands to create target and staging paths, in seconds")
+	stringVar(&config.RemoveTargetPathCmd, "removemountpathcmd", "Command to run for target path removal")
+	stringVar(&config.RemoveStagingPathCmd, "removestagingpathcmd", "Command to run for staging path removal")
+	intVar(&config.RemovePathCmdTimeout, "removepathcmdtimeout", "Timeout for the commands to remove target and staging paths, in seconds")
+	stringVar(&config.SecretsFile, "secrets", "CSI secrets file")
+	int64Var(&config.TestVolumeSize, "testvolumesize", "Base volume size used for provisioned volumes")
+	int64Var(&config.TestVolumeExpandSize, "testvolumeexpandsize", "Target size for expanded volumes")
+	stringVar(&config.TestVolumeParametersFile, "testvolumeparameters", "YAML file of volume parameters for provisioned volumes")
+	boolVar(&config.TestNodeVolumeAttachLimit, "testnodevolumeattachlimit", "Test node volume attach limit")
+	stringVar(&config.JUnitFile, "junitfile", "JUnit XML output file where test results will be written")
+
+	flag.Parse()
+	if *version {
 		fmt.Printf("Version = %s\n", VERSION)
-		return
+		os.Exit(0)
 	}
-	if len(config.Address) == 0 {
-		t.Fatalf("--%sendpoint must be provided with an CSI endpoint", prefix)
+	if config.Address == "" {
+		fmt.Printf("--%sendpoint must be provided with an CSI endpoint", prefix)
+		os.Exit(1)
 	}
-	sanity.Test(t, &config)
+	os.Exit(m.Run())
+}
+
+func TestSanity(t *testing.T) {
```

Comment:
I am wondering whether we need this special hack with `csi-sanity` being a `go test` program. It prevents using "go get" to build it. I might still change that as part of this PR...

Reply:
Yes, that was pretty simple.

Reply:
See https://github.com/kubernetes-csi/csi-test/pull/234.